### PR TITLE
fix: no switchover candidate found with member state "streaming" (#1992)

### LINKED ISSUE #1992: No switchover candidate found
Helllo,

we are encountering an error while cluster updating (any change in the configuration that needs pods restart), where the master is not correctly switched to the synchronous replica. We have 2 instances, one is master and one is a synchronous replica.

Here is log from the operator, the problem occurs every time:

> time="2022-08-10T11:39:52Z" level=info msg="reason: new statefulset containers's postgres (index 0) environment does not match the current one" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:39:52Z" level=debug msg="updating statefulset" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:39:53Z" level=debug msg="making GET http request: http://10.1.63.119:8008/config" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:05Z" level=debug msg="making GET http request: http://10.1.11.85:8008/patroni" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:05Z" level=debug msg="making GET http request: http://10.1.63.119:8008/patroni" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:05Z" level=debug msg="performing rolling update" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:05Z" level=info msg="there are 2 pods in the cluster to recreate" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:05Z" level=debug msg="subscribing to pod \"postgres-qplay/qplay-prod-0\"" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:27Z" level=info msg="pod \"postgres-qplay/qplay-prod-0\" has been recreated" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:27Z" level=debug msg="unsubscribing from pod \"postgres-qplay/qplay-prod-0\" events" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:27Z" level=debug msg="making GET http request: http://10.1.63.119:8008/cluster" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:27Z" level=warning msg="error while syncing cluster state: could not sync statefulsets: could not recreate pods: skipping switchover: no switchover candidate found" cluster-name=postgres-qplay/qplay-prod pkg=cluster
time="2022-08-10T11:40:27Z" level=error msg="could not sync cluster: could not sync statefulsets: could not recreate pods: skipping switchover: no switchover candidate found" cluster-name=postgres-qplay/qplay-prod pkg=controller worker=0
time="2022-08-10T11:40:27Z" level=info msg="received add event for already existing Postgres cluster" cluster-name=postgres-qplay/qplay-prod pkg=controller worker=0

Any idea what could be wrong? It could be definitely some wrong configuration (maybe three instances are the minimum?), but I can't figure out that it is.

### PR DESCRIPTION
**Pull request by bootc**
Fixes #1992

### CODE REVIEW

**Comment by hughcapet** [comment]
There is another possible state of a running replica - "in archive recovery"
https://github.com/zalando/patroni/commit/d46ca88e6bb4b4788ec86689026c2d2e4da68c79

**Comment by FxKu** [comment]
:+1:

File: pkg/cluster/pod.go [hunk 1]
```diff
@@ -503,7 +505,11 @@ func (c *Cluster) getSwitchoverCandidate(master *v1.Pod) (spec.NamespacedName, e
 	} else {
 		// in asynchronous mode find running replicas
 		for _, member := range members {
-			if PostgresRole(member.Role) != Leader && PostgresRole(member.Role) != StandbyLeader && member.State == "running" {
+			if PostgresRole(member.Role) == Leader || PostgresRole(member.Role) == StandbyLeader {
+				continue
+			}
+
+			if slices.Contains([]string{"running", "streaming", "in archive recovery"}, member.State) {
```
Comment (macedigital) [inline pkg/cluster/pod.go:512]: The next part of the code sorts candidates by smallest amount of lag. My only concern with merging this change might cause a switchover to a replica lagging behind significantly. Overall, this is still better than what we have now, though.

https://github.com/zalando/postgres-operator/blob/45489d3741b1b956c84e4779edbfd27a26970d9e/pkg/cluster/pod.go#L511-L516

**Review by macedigital — APPROVED** [review]
Let's merge this PR and make a new release for fixing the underlying issue.

File: pkg/cluster/pod_test.go [hunk 2]
```diff
@@ -75,6 +75,13 @@ func TestGetSwitchoverCandidate(t *testing.T) {
 			expectedCandidate: spec.NamespacedName{},
 			expectedError:     fmt.Errorf("no switchover candidate found"),
 		},
+		{
+			subtest:           "replicas with streaming state",
+			clusterJson:       `{"members": [{"name": "acid-test-cluster-0", "role": "leader", "state": "running", "api_url": "http://192.168.100.1:8008/patroni", "host": "192.168.100.1", "port": 5432, "timeline": 1}, {"name": "acid-test-cluster-1", "role": "replica", "state": "streaming", "api_url": "http://192.168.100.2:8008/patroni", "host": "192.168.100.2", "port": 5432, "timeline": 1, "lag": 5}, {"name": "acid-test-cluster-2", "role": "replica", "state": "streaming", "api_url": "http://192.168.100.3:8008/patroni", "host": "192.168.100.3", "port": 5432, "timeline": 1, "lag": 2}]}`,
```
Comment (macedigital) [inline pkg/cluster/pod_test.go:80]: We might want to add test cases to capture "in archive recovery" mode, eventually.

File: pkg/cluster/pod.go (hunk 1 above)
Comment (FxKu) [reply pkg/cluster/pod.go:512]: Patroni will prevent it if this is the case.

**Comment by hughcapet** [comment]
👍

**Pull request merged**